File: c/serial/serial.c
#include <stdio.h>

int main (int argc, char **argv)
{
   void datread(char *, void *, int , int );
   void pgmwrite(char *, void *, int , int );

   int i,j,k;
   int m = 600;
   int n = 450;
   int numit = 50000;
  
   float im[m+2][n+2], old[m+2][n+2], new[m+2][n+2];
   float buf[m][n];

   for (i = 0; i < m+2; i++)
     for (j = 0; j < n+2; j++)
       {
	  im[i][j] = 0.0;
	  old[i][j] = 0.0;
       }

   datread("edge600x450.dat", buf, m, n);

   for (i = 1; i <= m; i++)
     for (j = 1; j <= n; j++)
       {
	  im[i][j] = buf[i-1][j-1];
	  old[i][j] = buf[i-1][j-1];
       }

   for (k = 1; k <= numit; k++)
     {
	for (i = 1; i <= m; i++)
	  for (j = 1; j <= n; j++)
	    new[i][j] = 0.25*(old[i-1][j]+old[i+1][j]+old[i][j-1]+old[i][j+1]-
			      im[i][j]);
	for (i = 1; i <= m; i++)
	  for (j = 1; j <= n; j++)
	    old[i][j] = new[i][j];
	if (!(k % 1000)) printf("%d iterations done\n", k);
     }

   for (i = 1; i <= m; i++)
     for (j = 1; j <= n; j++)
       buf[i-1][j-1] = old[i][j];

   pgmwrite("image.pgm", buf, m, n);
    
   return 0;
}

File: fortran/serial/serial.f90
	program readwrite

	implicit none
	integer m,n,i,j,k,numit,nn,mm
	parameter (m = 600)
	parameter (n = 450)
	parameter (numit = 10000)
        character (len=32) :: filename
	
	real im(0:m+1,0:n+1),old(0:m+1,0:n+1),new(0:m+1,0:n+1)
	real buf(m,n)

        im(:,:)  = 0.e0
        old(:,:) = 0.e0	

        write(filename,'(a,I3,a,I3,a)'),"edge",m,"x",n,".dat"
	call datread(filename,buf,m,n)

        im(1:m,1:n) = buf(1:m,1:n)
        old(1:m,1:n) = buf(1:m,1:n)


	do k = 1, numit
          new(1:m,1:n) = 0.25e0*(old(0:m-1,1:n) + old(2:m+1,1:n) + old(1:m,0:n-1) + old(1:m,2:n+1) - im(1:m,1:n))
	  old(:,:)=new(:,:)
          if ( mod(k,1000) == 0 ) then
          write(*,*) 'Iteration # ',k
          end if
	end do
	
	buf(1:m,1:n)=old(1:m,1:n)
	
	call pgmwrite('image.pgm',buf,m,n)
	
	end

